addons/binding/org.openhab.binding.russound/src/main/java/org/openhab/binding/russound/internal/discovery/RioSystemDeviceDiscoveryService.java

```diff
@@ -0,0 +1,261 @@
+    private static final Pattern RSP_SRCNOTIFICATION = Pattern.compile("(?i)^[SN] S\\[(\\d+)\\]\\.(\\w+)=\"(.*)\"$");
+
+    /** Pattern to identify zone notifications */
+    private static final Pattern RSP_ZONENOTIFICATION = Pattern
+            .compile("(?i)^[SN] C\\[(\\d+)\\]\\.Z\\[(\\d+)\\]\\.(\\w+)=\"(.*)\"$");
+
+    /**
+     * The {@link SocketSession} that will be used to scan the device
+     */
+    private SocketSession session;
+
+    /**
+     * The {@link WaitingSessionListener} to the {@link #session} to receive/process responses
+     */
+    private WaitingSessionListener listener;
+
+    /**
+     * Create the discovery service from the {@link RioSystemHandler}
+     *
+     * @param sysHandler a non-null {@link RioSystemHandler}
+     * @throws IllegalArgumentException if sysHandler is null
+     */
+    public RioSystemDeviceDiscoveryService(RioSystemHandler sysHandler) {
+        super(RussoundHandlerFactory.SUPPORTED_THING_TYPES_UIDS, 30, false);
+
+        if (sysHandler == null) {
+            throw new IllegalArgumentException("sysHandler can't be null");
+        }
+        this.sysHandler = sysHandler;
+    }
+
+    /**
+     * Activates this discovery service. Simply registers this with
+     * {@link RioSystemHandler#registerDiscoveryService(RioSystemDeviceDiscoveryService)}
+     */
+    public void activate() {
+        sysHandler.registerDiscoveryService(this);
+    }
+
+    /**
+     * Deactivates the scan - will disconnect the session and remove the {@link #listener}
+     */
+    @Override
+    public void deactivate() {
+        if (session != null) {
+            try {
+                session.disconnect();
+            } catch (IOException e) {
+                // ignore
+            }
+            session.removeListener(listener);
+            session = null;
+            listener = null;
+        }
+    }
+
+    /**
+     * Overridden to do nothing - {@link #scanDevice()} is called by {@link RioSystemHandler} instead
+     */
+    @Override
+    protected void startScan() {
+        // do nothing - started by RioSystemHandler
+    }
+
+    /**
+     * Starts a device scan. This will connect to the device and discover the controllers/sources/zones attached to the
+     * device and then disconnect via {@link #deactivate()}
+     */
+    public void scanDevice() {
+        try {
+            final String ipAddress = sysHandler.getRioConfig().getIpAddress();
+            session = new SocketChannelSession(ipAddress, RioConstants.RioPort);
+            listener = new WaitingSessionListener();
+            session.addListener(listener);
+
+            try {
+                logger.debug("Starting scan of RIO device at {}", ipAddress);
+                session.connect();
+                discoverControllers();
+                discoverSources();
+            } catch (IOException e) {
+                logger.warn("Trying to scan device but couldn't connect: {}", e.getMessage(), e);
```

Comment:
that might be expected, it the device is not reachable, right? So better reduce this to debug.